### ISSUE TITLE
Cannot load string streams with more than one channel

### ISSUE THREAD

**Issue by lgarbar** [cbrnr/XDF.jl#10]
The read_xdf function seems to work without problem for some .xdf files and not others. Listed below is the full error message from my Pluto notebook. Importantly, for the files that fail and the ones that succeed don't seem to have any reliable difference in size (given that they're from the same task performed by different participants) and compile the same number of streams and headers. Any insight would be appreciated.

Error message:
MethodError: no method matching read(::IOStream, ::Nothing)
Closest candidates are:
read(::IOStream)
@ Base iostream.jl:519
read(::IOStream, !Matched::Type{Float64})
@ Base iostream.jl:422
read(::IOStream, !Matched::Type{Float32})
@ Base iostream.jl:421
...
(::Main.var"workspace#18".myXDF.var"#1#3"{Dict{Int64, Int64}, Dict{Int64, Any}})(::IOStream)@my_XDF.jl:130
var"#open#424"(::Base.Pairs{Symbol, Union{}, Tuple{}, @NamedTuple{}}, ::typeof(open), ::Main.var"workspace#18".myXDF.var"#1#3"{Dict{Int64, Int64}, Dict{Int64, Any}}, ::String)@io.jl:396
open(::Function, ::String)@io.jl:393
read_xdf(::String, ::Bool)@my_XDF.jl:36
read_xdf(::String)@my_XDF.jl:33
top-level scope@Local: 8

**Comment by cbrnr** [comment]
Can you provide a minimal reproducible example with a file that does not work (i.e. no notebook) and then post the stack trace?

Did you try if e.g. pyxdf can read your files?

**Comment by lgarbar** [comment]
Hello,

The datasets that I've seen fail are participant data, so I won't be able to share any of those data and certainly not through a public forum. Perhaps I'll check with my supervisor if it's acceptable to share fully de-identified and encrypted data with you privately so you can take a look. Otherwise, the issue appears for some files and not others, so I can try to record some dummy data to see if I can reproduce the error and share that instead.

Otherwise, I have had no issues with pyxdf, but it is significantly slower in my experience than unpacking with Julia, so we're looking to potentially make a switch.

**Comment by cbrnr** [comment]
It would be great if you recorded some dummy data to reproduce the problem. Otherwise, it will be almost impossible to determine what the issue is.

**Comment by lgarbar** [comment]
How might you suggest sharing the data file? The dummy data I have here is >100mb, so I can't attach it here nor push it to a repo.

**Comment by cbrnr** [comment]
Do you have Google Drive, OneDrive, Dropbox, etc.? If not, I can share an upload link.

**Comment by lgarbar** [comment]
Sorry, I guess I meant if you had a preference. Here's a [onedrive link](https://nysemail-my.sharepoint.com/personal/luis_garcia-barnett_nki_rfmh_org/_layouts/15/onedrive.aspx?csf=1&web=1&e=5XNeGU&cid=7d5ef9e7%2D8996%2D4898%2Da199%2D52494c5a19ba&id=%2Fpersonal%2Fluis%5Fgarcia%2Dbarnett%5Fnki%5Frfmh%5Forg%2FDocuments%2Fxdf%5Fjl%5Ftroubleshooting&FolderCTID=0x01200089796E252DA3F141A60A395E9BE5C59E&view=0) with the data, jl script, and stack trace listed in a txt file

**Comment by cbrnr** [comment]
This link doesn't work, it seems to be restricted for internal use.

**Comment by lgarbar** [comment]
Yeah, makes sense. Try this [link](https://drive.google.com/file/d/1pi4mh2vfhdxwjRQMgjtgCCxiX7rvY3hb/view?usp=share_link)

**Comment by lgarbar** [comment]
Makes sense. Try this [link](https://drive.google.com/drive/folders/1dH3vrptNLWYBA3B_ofUCvqCYQkkJRTHl?usp=share_link)

**Comment by cbrnr** [comment]
Yes, that link worked!

**Comment by cbrnr** [comment]
Something's wrong when reading chunk 9 (stream ID 5, chunk type 3). Somehow, there's a length of 17 bytes where there should be either 1, 4, or 8 bytes. This could be due to two reasons (I can think of): (1) it's a bug or (2) string streams are not supported yet. I'm still trying to figure it out...

**Comment by lgarbar** [comment]
Hmm that's interesting. I imagine the issue itself isn't with the datatype being strings since I'm able to unpack our other data of the same format and with string streams.

**Comment by cbrnr** [comment]
It definitely has to do with your data containing a regularly sampled string stream. I have never seen this in the wild (and hence I don't have any testing data), so I did not implement support for it (yet). Of course, since these are perfectly valid in XDF, these should be supported, but someone will need to implement it. Currently, I don't have the time to do that, but I'll at least mention this limitation in the README.

**Comment by lgarbar** [comment]
Interesting. I'll try collecting a set of dummy data that can be successfully read to see if you can identify any systemic differences between the two datasets.

**Comment by abcsds** [comment]
Oh, I'm just reading this now. Hi @lgarbar I implemented the string markers and can help you debug this if you're still interested. 

Let me tell you how I'm picking up this issue: I downloaded the file you provide, a 481.1 Mb xdf. A stream, FaceVideo, with 25344 channels of type float32. Could you please confirm this is the file? I have a sha256 of "836ef458ba94bfa95a5fc1f42093cf671e012632435e5159cb12853335962dbe".

Upon reading chunk by chunk with the debugger I found no issue, but integrating it to tests I get weird error messages about the `read()` function call in [L36](https://github.com/cbrnr/XDF.jl/blob/18e055e5e2c191197c3353f508bf5c0e0e8dac82/src/XDF.jl#L36)

```
The function `read` exists, but no method is defined for this combination of argument types.

Closest candidates are:
  read(::Any, ::Type{Union{}}, Any...; kwargs...)
   @ Base io.jl:236
  read(::IO, ::Type{Char})
   @ Base io.jl:962
  read(::IO, ::Type{T}) where T<:Enum
   @ Base Enums.jl:22
  ...
```

These type of errors are common on my linux ditsros, and I usually solve them with a container. I'll test again the file you sent, but It would be of much help if you can tell us more about the origin of the data: was it created by a provider? the LibLSL libraries? Is it working if you use the [pyxdf](https://github.com/xdf-modules/pyxdf)? 

Thanks for letting us know about the issue.


@cbrnr It might be better to pause #11 if this is a bug in the code. I hadn't seen this issue. Next time please feel free to tag me about bugs, specially if you think they might be in code I implemented. Still, I don't see string streams in the file, just the float32 stream. Is it the same file you revised?

**Comment by cbrnr** [comment]
@abcsds could it be because you are assuming that there is only one channel? Stream 5 has 22 channels and dtype string...

**Comment by cbrnr** [comment]
@abcsds There is a much smaller test file with a string stream consisting of 2 channels sampled at 1000Hz (maybe useful for testing) linked in this issue:

https://github.com/xdf-modules/libxdf/issues/19

**Comment by abcsds** [comment]
The file from https://github.com/xdf-modules/libxdf/issues/19 reads fine with XDF.jl. The problem is unrelated. 

The file provided has chunk 522 start with chunk bytesize 17, which is not a valid size for XDF (https://github.com/sccn/xdf/wiki/Specifications#chunk). First tests seem to indicate the error is in how the file was written. @lgarbar could you tell us more about it? Was it generated with labrecorder?

Next steps could be checking against pyxdf, but I think it would still be useful to have a minimal example that reproduces the error.

On my dev branch (https://github.com/abcsds/XDF.jl/tree/dev) I've added a simple error catch, which should point users in the right direction.

**Comment by cbrnr** [comment]
FYI pyxdf loads the file without problems.

```
[{'channel_count': 1,
  'channel_format': 'string',
  'created_at': '100139.2259153700',
  'hostname': 'nkirs-Precision-3650-Tower',
  'name': 'StimLabels',
  'nominal_srate': 0.0,
  'session_id': 'default',
  'source_id': 'markers-nkirs-Precision-3650-Tower',
  'stream_id': 6,
  'type': 'Markers',
  'uid': 'bffc2f2f-42c2-417c-bd48-98a97ed9064c'},
 {'channel_count': 1,
  'channel_format': 'string',
  'created_at': '112942.2599068000',
  'hostname': 'MobiLabEEG',
  'name': 'BrainVision RDA Markers',
  'nominal_srate': 0.0,
  'session_id': 'default',
  'source_id': 'RDA 10.10.10.5:51244 Marker',
  'stream_id': 4,
  'type': 'Markers',
  'uid': '88cac138-d742-4644-9e45-fc50f729b4bb'},
 {'channel_count': 5,
  'channel_format': 'float32',
  'created_at': '113253.2100808',
  'hostname': 'DESKTOP-KMDE17V',
  'name': 'OpenSignals',
  'nominal_srate': 1000.0,
  'session_id': 'default',
  'source_id': 'OpenSignals',
  'stream_id': 1,
  'type': '00:07:80:8C:0A:53',
  'uid': '22aa6dc6-20dd-4bfd-abd0-8af7467ce66b'},
 {'channel_count': 25344,
  'channel_format': 'float32',
  'created_at': '100120.0217223920',
  'hostname': 'nkirs-Precision-3650-Tower',
  'name': 'FaceVideo',
  'nominal_srate': 30.0,
  'session_id': 'default',
  'source_id': 'facevid-nkirs-Precision-3650-Tower',
  'stream_id': 2,
  'type': 'Video',
  'uid': '6f59d30f-f4cb-4769-a54b-d9d935e7adad'},
 {'channel_count': 22,
  'channel_format': 'string',
  'created_at': '7648.225881600000',
  'hostname': 'EYETrack',
  'name': 'Argus_Eye_Tracker',
  'nominal_srate': 180.0,
  'session_id': 'default',
  'source_id': 'argus-EYETrack',
  'stream_id': 5,
  'type': 'Gaze',
  'uid': '8d114fd8-df7a-4a18-830a-004399bbfdae'},
 {'channel_count': 65,
  'channel_format': 'float32',
  'created_at': '112942.2581061000',
  'hostname': 'MobiLabEEG',
  'name': 'BrainVision RDA',
  'nominal_srate': 1000.0,
  'session_id': 'default',
  'source_id': 'RDA 10.10.10.5:51244',
  'stream_id': 3,
  'type': 'EEG',
  'uid': '76d828ea-c40b-4254-83e4-38b67f242d97'},
 {'channel_count': 1,
  'channel_format': 'int16',
  'created_at': '113267.6849361000',
  'hostname': 'DESKTOP-KMDE17V',
  'name': 'Audio',
  'nominal_srate': 44100.0,
  'session_id': 'default',
  'source_id': 'audio-DESKTOP-KMDE17V',
  'stream_id': 7,
  'type': 'AudioCapture',
  'uid': 'f8b8e01a-8fae-4346-b94c-94717a75afb0'}]
```

**Comment by abcsds** [comment]
Found it, thanks. Tests on https://github.com/xdf-modules/libxdf/issues/19 weren't actually passing; it had the same channel entry on both channels. I've fixed the multi-channel string marker reading and updated the test for the minimal example. The file provided by @lgarbar opens too.
